relayer-cli/src/commands/misbehaviour.rs

```diff
@@ -0,0 +1,106 @@
+use ibc::ics02_client::height::Height;
+use ibc::ics24_host::identifier::ChainId;
+use ibc_relayer::config::StoreConfig;
+use ibc_relayer::foreign_client::ForeignClient;
+
+use crate::application::CliApp;
+use crate::commands::cli_utils::{spawn_chain_runtime, SpawnOptions};
+use crate::conclude::Output;
+use crate::prelude::*;
+
+#[derive(Clone, Command, Debug, Options)]
+pub struct MisbehaviourCmd {
+    #[options(
+        free,
+        required,
+        help = "identifier of the chain where client updates are monitored for misbehaviour"
+    )]
+    chain_id: ChainId,
+}
+
+impl Runnable for MisbehaviourCmd {
+    fn run(&self) {
+        let config = app_config();
+
+        let res = monitor_misbehaviour(&self.chain_id, &config);
+        match res {
+            Ok(()) => Output::success(()).exit(),
+            Err(e) => Output::error(format!("{}", e)).exit(),
+        }
+    }
+}
+
+pub fn monitor_misbehaviour(
+    chain_id: &ChainId,
+    config: &config::Reader<CliApp>,
+) -> Result<(), BoxError> {
+    let spawn_options = SpawnOptions::override_store_config(StoreConfig::memory());
+
+    let chain = spawn_chain_runtime(&spawn_options, &config, chain_id).unwrap();
+
+    let subscription = chain.subscribe()?;
+
+    while let Ok(event_batch) = subscription.recv() {
+        for event in event_batch.events.iter() {
+            match event {
+                IbcEvent::UpdateClient(update) => {
+                    dbg!(update);
+                    // 1 - make sure client is supported and then create the counterparty chain handle
+                    let client_id = update.client_id();
+                    let client_state = chain.query_client_state(client_id, Height::zero()).unwrap();
+                    let tm_client_state =
+                        downcast!(client_state => AnyClientState::Tendermint).unwrap();
+                    let counterparty_chain = spawn_chain_runtime(
+                        &spawn_options,
+                        &config,
+                        &ChainId::from_str(tm_client_state.chain_id.as_str()).unwrap(),
+                    )
+                    .unwrap();
+
+                    // 2 - restore the foreign client
+                    let client = ForeignClient::restore_client(
+                        chain.clone(),
+                        counterparty_chain.clone(),
+                        client_id,
+                    );
+
+                    // 3 - decode light block from the update event
+                    // TODO - code below may change depending on the header representation in the new update event
+                    // Currently this doesn't work as `update.header` has a weird format
+                    if update.client_type() == ClientType::Tendermint {
+                        let chain_header = Any {
+                            type_url: TENDERMINT_HEADER_TYPE_URL.to_string(),
+                            value: Vec::from(update.header.as_bytes()),
```

Comment:
this seems a little odd to me. The header bytes should contain the type_url? This is the definition in the [SDK](https://github.com/cosmos/cosmos-sdk/blob/a193522f7e54cd747d549bc4adb83618aac223af/codec/types/any.go#L77). I don't know rust, but it looks like you are creating an `Any` by passing in the type url using a constant and then trying to decode the encoded any into the value of the `Any`?

```
Any{
    type: tendermint,
    value: Any{
        type: tendermint,
        value: ibc.header{}
        }
  }
  ```

Reply:
Here is how the header is being encoded in the [SDK](https://github.com/cosmos/cosmos-sdk/blob/a193522f7e54cd747d549bc4adb83618aac223af/codec/proto_codec.go#L171)